pandas/core/reshape/reshape.py

```diff
@@ -876,3 +881,110 @@ def _reorder_for_extension_array_stack(
     #  c0r1, c1r1, c2r1, ...]
     idx = np.arange(n_rows * n_columns).reshape(n_columns, n_rows).T.ravel()
     return arr.take(idx)
+
+
+def stack_v3(frame: DataFrame, level: list[int]) -> Series | DataFrame:
+    if frame.columns.nunique() != len(frame.columns):
+        raise ValueError("Columns with duplicate values are not supported in stack")
+
+    # If we need to drop `level` from columns, it needs to be in descending order
+    drop_levnums = sorted(level)[::-1]
+    stack_cols = frame.columns._drop_level_numbers(
+        [k for k in range(frame.columns.nlevels) if k not in level][::-1]
+    )
+    if len(level) > 1:
+        # Arrange columns in the order we want to take them, e.g. level=[2, 0, 1]
+        sorter = np.argsort(level)
+        ordered_stack_cols = stack_cols._reorder_ilevels(sorter)
+    else:
+        ordered_stack_cols = stack_cols
+
+    stack_cols_unique = stack_cols.unique()
+    ordered_stack_cols_unique = ordered_stack_cols.unique()
+
+    # Grab data for each unique index to be stacked
+    buf = []
+    for idx in stack_cols_unique:
+        if len(frame.columns) == 1:
+            data = frame.copy()
+        else:
+            # Take the data from frame corresponding to this idx value
+            if not isinstance(idx, tuple):
+                idx = (idx,)
+            gen = iter(idx)
+            column_indexer = tuple(
+                next(gen) if k in level else slice(None)
+                for k in range(frame.columns.nlevels)
+            )
+            data = frame.loc[:, column_indexer]
```

Comment:
Does `column_indexer` need to be concerned about duplicate labels?

Reply:
Not 100% this is what you're asking, but the column_indexer is accessing different values across levels of the columns. E.g.

<details>

```
columns = pd.MultiIndex(
    levels=[list("xy"), list("xy")],
    codes=[[0, 1], [0, 1]],
)
df = pd.DataFrame([[0, 1], [2, 3]], columns=columns)
print(df)
#    x  y
#    x  y
# 0  0  1
# 1  2  3

print(df.loc[:, ("x", "x")])
# 0    0
# 1    2
# Name: (x, x), dtype: int64
```

</details>

Reply:
Right, but if the levels were `[list("xyx"), list("xyx")]` would it be correct to select the multiple `("x", "x")` levels?

Reply:
On L886 we raise when columns have duplicates (as discussed in #53761).